modules/tinymce/src/core/test/ts/browser/util/ImageUploaderTest.ts

```diff
@@ -0,0 +1,66 @@
+import { Assert, UnitTest } from '@ephox/bedrock-client';
+import { ApiChains, TinyLoader } from '@ephox/mcagar';
+import { Arr } from '@ephox/katamari';
+import ImageUploader, { UploadResult } from 'tinymce/core/api/util/ImageUploader';
+
+import { BlobCache, BlobInfo } from 'tinymce/core/api/file/BlobCache';
+
+import Theme from 'tinymce/themes/silver/Theme';
+
+UnitTest.asynctest('browser.tinymce.core.util.ImageUploaderTest', (success, failure) => {
+
+  Theme();
+
+  TinyLoader.setupLight((editor, onSuccess, onFailure) => {
+    const cache = BlobCache();
+    const blob = new Blob([ JSON.stringify({ hello: 'world' }) ]);
+
+    const image1 = cache.create({ blob, base64: 'test' });
+
+    const cUploadImages = (images: BlobInfo[], openNotification: boolean) => Chain.async((_value, next, die) => {
+      const uploader = ImageUploader(editor);
+      uploader.upload(images, openNotification).then(next, die);
+    });
+
+    const cAssertUploadResultSuccess = (expectedLength: number, expectedUrl: string) => Chain.op((uploadResults: UploadResult[]) => {
+      Assert.eq('Test', expectedLength, uploadResults.length);
+      Arr.each(uploadResults, (uploadResult) => {
+        Assert.eq('Url is Image.png', expectedUrl, uploadResult.url);
+        Assert.eq('Upload result status is true upon success', true, uploadResult.status);
+      });
+    });
+
+    const cAssertUploadResultFailure = (expectedLength: number, expectedUrl?: string, errorMsg?: string) => Chain.op((uploadResults: UploadResult[]) => {
+      Assert.eq('Test', expectedLength, uploadResults.length);
+      Arr.each( uploadResults, (uploadResult) => {
+        Assert.eq('Url is empty string upon failure', expectedUrl, uploadResult.url);
+        Assert.eq('Upload result status is false upon failure', false, uploadResult.status);
+        Assert.eq('Upload result error message matches failure message', errorMsg, uploadResult.error.message);
+      });
+    });
+
+    Pipeline.async({}, [
+      Log.chainsAsStep('TINY-4601', 'Image upload success', [
+        Chain.inject(editor),
+        ApiChains.cSetSetting('images_upload_handler', (_blobInfo: BlobInfo[], success) => success('https://tiny.cloud/image.png')),
+        cUploadImages([ image1 ], true),
+        cAssertUploadResultSuccess(1, 'https://tiny.cloud/image.png')
+      ]),
+      Log.chainsAsStep('TINY-4601', 'Image upload failure', [
+        Chain.inject(editor),
+        ApiChains.cSetSetting('images_upload_handler', (_blobInfo: BlobInfo[], success, failure) => failure('Error msg')),
+        cUploadImages([ image1 ], true),
+        cAssertUploadResultFailure(1, '', 'Error msg')
+      ]),
+      Log.chainsAsStep('TINY-4601', 'Image upload failure for empty array', [
+        Chain.inject(editor),
+        ApiChains.cSetSetting('images_upload_handler', (_blobInfo: BlobInfo[], success, failure) => failure('Error msg')),
+        cUploadImages([ ], true),
+        cAssertUploadResultFailure(0)
+      ])
```

Comment:
Might be worth adding some tests for multiple images. 

_Note:_ You'll need an `image2` so you can't just batch this suggestion in
```suggestion
      ]),
      Log.chainsAsStep('TINY-4601', 'Multiple image upload success', [
        Chain.inject(editor),
        ApiChains.cSetSetting('images_upload_handler', (_blobInfo: BlobInfo, success) => success('https://tiny.cloud/image.png')),
        cUploadImages([ image1, image2 ], true),
        cAssertUploadResultSuccess(2, 'https://tiny.cloud/image.png')
      ]),
      Log.chainsAsStep('TINY-4601', 'Multiple image upload failure', [
        Chain.inject(editor),
        ApiChains.cSetSetting('images_upload_handler', (_blobInfo: BlobInfo, success, failure) => failure('Error msg')),
        cUploadImages([ image1, image2 ], true),
        cAssertUploadResultFailure(2, '', 'Error msg')
      ]),
```